src/config_design.config

```diff
@@ -0,0 +1,126 @@
+%% Cloudant_benchmark config changes design review 
+
+%% ----------------------------------------------------------------------------
+%% CONFIGURATION TEMPLATES: (Fogbugz 62351)
+%% ----------------------------------------------------------------------------
+
+{config_templates, ["basho_bench", "cloudant_http", "http_driver"]}.
+
+%% IMPLEMENTATION
+%% Optional configuration template files to predefine parameters/preset default values
+%% Each record would contain something like (Name,Value,ValidatorFun,Description,?MODULE) 
+%%
+%% Config template files could be nested, 
+%%    would allow us to validate-on-load incoming parameters that match
+%%    warn about unknown config params to catch typos
+%%    used to auto-generate a key-value only config file for local config file use, 
+%%    would help dumping ALL config params+doc to run log, not just those provided by user
+%% 
+%% basho_bench_config:load
+%%    would pre-process terms instead of copying directly to application environmant
+%%    
+%%    if "config_templates" call then loads those config template files, 
+%%        saving parameter records into common ETS table
+%%        presetting parameter value defaults into environment
+%%
+%%    else if any templates have been defined,
+%%        checks term against known parameter records
+%%        if known then checks value with ValidatorFun before saving to environment
+%%        else warns about unknown parameters (potential typos) but accepts
+
+%% ----------------------------------------------------------------------------
+%% PER-OPERATIONS CONFIGURATION: https://github.com/cloudant/perf/issues/23 and 29
+%% ----------------------------------------------------------------------------
+%%
+%% Operations would change to allow optional per-operation keygen and valgen
+
+{operations, [
+   %% basho_bench_worker and _stats expect two values - Name, Count 
+   {get, 1},
+
+   %% Using alternate {op,[list]} format keeps 2-tuples so either format could be used
+   %% QUESTION: Do we need ONLY keygen or the OPTION of BOTH keygen and valgen ? 
+   %%    Would we ever need MORE per-op options? 
+
+   {put, [{count,1},  % Alternate count in more complex format, modify existing lookups
+          {key_generator, {int_to_str, {partitioned_sequential_int, 100000}}},
+          {value_generator, {function,http_couchdb_driver,make_json_doc,[]}}
+	 ]}
+   ]}.
+
+%% IMPLEMENTATION:
+%% basho_bench_worker:init 
+%%    turns per-op config into new Keygen/ValueGen entities
+%%    and stores into worker State as per-op-config list
+%%        per_op_config =[{put, [{keygen, KeyGen}, {valgen, ValGen}, ...]}, ... ]
+%% basho_bench_worker:worker_next_op2 
+%%    checks per_op_config, uses as run call arguments,
+%%       potentially stores onto State as current keygen/valgen
+
```

Comment:
The problem I am realizing now is that its easy to access the workers/worker_type config information in basho_bench_worker:init and save it into the worker State, all that gets passed in the Driver:new call is the worker ID not the developing state information or even the workerType to be able to lookup the information from config again. 

Similarly the run method only contains the DriverState created by Driver:new() so again there isn't enough being passed down to complete this "last mile" to per-worker config information. 

Few thoughts
1. Kludge - Work outside the stack. Store the worker State into a table so it can be looked up by the Id provided to new() and then have the individual application store the Id into its own DriverState for run to access it later if desired. UGLY

2. Redefine the new() and run() signatures to pass either the overall State, the localConfig inside it, etc. 

3. Push the LocalConfig into the process environment using get(K) and put(K,V) rather than storing it in its own State-based list. Then adjust the new basho_bench_worker:config_get function to use the process environment OR basho_bench_config dependning on what it finds. Sounds like the most sensible so I will experiment with this.